specs/WebResourceResponseReceived.md

```diff
@@ -0,0 +1,232 @@
+    interface ICoreWebView2WebResourceResponseReceivedEventArgsPopulateResponseContentCompletedHandler : IUnknown
+    {
+        /// Called to provide the implementer with the completion status
+        /// of the corresponding asynchronous method call.
+        HRESULT Invoke([in] HRESULT errorCode);
+    }
+
+    /// Event args for the WebResourceResponseReceived event. Will contain the
+    /// request as it was sent and the response as it was received.
+    /// Note: To get the response content stream, first call PopulateResponseContent
+    /// and wait for the async call to complete, otherwise the content stream object
+    /// returned will be null.
+    interface ICoreWebView2WebResourceResponseReceivedEventArgs : IUnknown
+    {
+        /// Web resource request object. Any modifications to this object will be ignored.
+        [propget] HRESULT Request([out, retval] ICoreWebView2WebResourceRequest** request);
+        /// Web resource response object. Any modifications to this object
+        /// will be ignored.
+        [propget] HRESULT Response([out, retval] ICoreWebView2WebResourceResponse** response);
+
+        /// Async method to ensure that the Content property of the response contains the actual response body content.
+        /// If this method is being called again before a first call has completed, all handlers are invoked at the same time.
+        /// If this method is being called after a first call has completed, the handler is invoked immediately.
+        HRESULT PopulateResponseContent(ICoreWebView2WebResourceResponseReceivedEventArgsPopulateResponseContentCompletedHandler* handler);
+    }
+}
+```
+
+## WinRT
+```c#
+namespace Microsoft.Web.WebView2.Core
+{
+    runtimeclass CoreWebView2
+    {
+        // ...
+
+        /// WebResourceResponseReceived event is raised after the WebView has received and processed the response for a WebResource request.
+        /// The event args include the WebResourceRequest as committed and the WebResourceResponse received,
+        /// including any additional headers added by the network stack that were not be included as part of
+        /// the associated WebResourceRequested event, such as Authentication headers.
+        event Windows.Foundation.TypedEventHandler<CoreWebView2, CoreWebView2WebResourceResponseReceivedEventArgs> WebResourceResponseReceived;
+    }
+
+    /// Event args for the WebResourceResponseReceived event.
+    /// Note: To get the response content stream, first call PopulateResponseContentAsync and
+    /// wait for the call to complete, otherwise the content stream object returned will be null.
+    runtimeclass CoreWebView2WebResourceResponseReceivedEventArgs
+    {
+        /// Web resource request object.
+        /// Any modifications to this object will be ignored.
+        CoreWebView2WebResourceRequest Request { get; };
+        /// Web resource response object.
+        /// Any modifications to this object will be ignored.
```

Comment:
Why is it a mutable type? Again it looks like we're trying to avoid creating a new type and reusing a type that isn't appropriate.

Reply:
I'm not sure if I'm picking the correct .NET HTTP type, but at least [System.Net.Http.HttpClient.PostAsync](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclient.postasync?view=netcore-3.1) asynchronously returns an instance of HttpResponse which has setters for most of its properties including its Content property. That is, in this very similar case the .NET http classes work similarly to what we're proposing.

Is copying a type to generate a non-mutable version a common pattern? Is there a good name suffix to use when following this pattern?

Reply:
IVector and IVectorView are basically the same example. But I think another case is [AllJoynAboutData](https://docs.microsoft.com/uwp/api/Windows.Devices.AllJoyn.AllJoynAboutData) and [AllJoynAboutDataView](https://docs.microsoft.com/uwp/api/Windows.Devices.AllJoyn.AllJoynAboutDataView).  [DataPackage](https://docs.microsoft.com/uwp/api/Windows.ApplicationModel.DataTransfer.DataPackage) and [DataPackageView](https://docs.microsoft.com/uwp/api/Windows.ApplicationModel.DataTransfer.DataPackageView). (I'm not sure if the View naming pattern is required, I was just searching for "View" classes.)

The goal is that IntelliSense tells you to do the write thing, and if that fails the compiler. But you shouldn't have to wait until runtime to find out you did something that's not supported.

Reply:
(a) WebResourceRequested - The response type allows the caller to generate a response instead of the WebView2 going to the wire and getting the actual response from the HTTP server.
(b) WebResourceResponseReceived - The response type allows the caller to view the actual response from the HTTP server.

Aspects of (b) different from (a):
 * immutable: view, readOnly, immutable
 * the body isn't available immediately: partial, inProgress
 * its the actual response from the server: ...?

CoreWebView2WebResourceResponseView
CoreWebView2WebResourceResponsePartial